3scale/porta
# 🚧 THREESCALE-7132: Add support for STS authentication for S3 bucket used in 3Scale

**What this PR does / why we need it**:

This PR adds a patch for Paperclip so that we're able to pass AWS `session_token`s to it. The idea is to support the use of temporary AWS tokens to access our S3 buckets.

**Which issue(s) this PR fixes** 

Implements https://issues.redhat.com/browse/THREESCALE-7132.

**Verification steps** 

Testing in local dev env:
1. Add valid credentials to `config/amazon_s3.yml`;
2. Perform a request using Paperclip;
          a.  Could be a file upload or a check if a file exists;
3. Assert that the `session_token` is used on the request;

**Special notes for your reviewer**:

This is the first draft of what we could use to support STS without migrating from Paperclip to ActiveStorage;
We still have to better understand how this `session_token` will be present to the Porta application, which most probably will result in further development.

Furthermore, I'd like to share a couple lines of code used to test AWS session tokens:

```ruby
# Use access_key_id, secret_access_key and region to instantiate an Aws::STS::Client
# and get the temporary session_token
sts = Aws::STS::Client.new(CMS::S3.credentials.merge(region: CMS::S3.region))
session = sts.get_session_token(duration_seconds: 15*60)
# Use the access_key_id, secret_access_key and temporary session_token (and the region again...)
# to access S3.
# NOTE: This "session" variable includes the access_key_id and secret_access_key
s3 = Aws::S3::Client.new(credentials: session, region: CMS::S3.region)
# Call whatever S3 api just to test if it's working properly
s3.get_bucket_tagging(bucket: 'some_bucket')
#=> #<struct Aws::S3::Types::GetBucketTaggingOutput...

# Also, Paperclip instantiates an Aws::S3::Resource to perform calls. This piece of code tests that a session_token is usable:
resource = Aws::S3::Resource.new(
  access_key_id: session.credentials.access_key_id,
  secret_access_key: session.credentials.secret_access_key,
  region: CMS::S3.region,
  session_token: session.credentials.session_token
)
resource.client.get_bucket_tagging(bucket: 'edge-flexo.3scale.net.3scale.net')

```
